pennylane/kernels/cost_functions.py

```diff
@@ -0,0 +1,135 @@
+# Copyright
+
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+
+#     http://www.apache.org/licenses/LICENSE-2.0
+
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+This file contains functionalities for kernel related costs.
+See `here <https://www.doi.org/10.1007/s10462-012-9369-4>`_ for a review.
+"""
+from pennylane import numpy as np
+from ..utils import frobenius_inner_product
+
+
+def square_kernel_matrix(X, kernel, assume_normalized_kernel=False):
+    """Computes the square matrix of pairwise kernel values for a given dataset.
+
+    Args:
+        X (list[datapoint]): List of datapoints
+        kernel ((datapoint, datapoint) -> float): Kernel function that maps datapoints to kernel value.
+        assume_normalized_kernel (bool, optional): Assume that the kernel is normalized, i.e.
+            that when both arguments are the same datapoint the kernel evaluates to 1. Defaults to False.
+
+    Returns:
+        array[float]: The square matrix of kernel values.
+    """
+    N = len(X)
+    matrix = [0] * N ** 2
+
+    for i in range(N):
+        for j in range(i, N):
+            if assume_normalized_kernel and i == j:
+                matrix[N * i + j] = 1.0
+            else:
+                matrix[N * i + j] = kernel(X[i], X[j])
+                matrix[N * j + i] = matrix[N * i + j]
+
+    return np.array(matrix).reshape((N, N))
+
+
+def kernel_matrix(X1, X2, kernel):
+    """Kernel polarization of a given kernel function.
+
+    Args:
+        X1 (list[datapoint]): List of datapoints (first argument)
+        X2 (list[datapoint]): List of datapoints (second argument)
+        kernel ((datapoint, datapoint) -> float): Kernel function that maps datapoints to kernel value.
```

Comment:
I really like the type hints here!

Reply:
Actually, it makes the docstrings nicely readable, so I changed my mind to be in favour of `datapoint` :)

Reply:
Perhaps something like?

```
kernel (function: (datapoint, datapoint) -> float)
```

Although, I don't know if this syntax is a valid type hint any more!

Reply:
Isn't the fact that it is a function already contained in the arrow?

Reply:
Let's leave it as is, I think this is clear!